currency-rates/internal/clients/chain_test.go

```diff
@@ -0,0 +1,87 @@
+package clients_test
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	"github.com/RinaDish/currency-rates/internal/clients"
+	"github.com/RinaDish/currency-rates/internal/clients/mocks"
+	"github.com/RinaDish/currency-rates/tools"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	"go.uber.org/zap"
+)
+
+var expectedSuccessRate = float64(10.0)
+var expectedFailRate = float64(0.0)
+
+type ChainTestSuite struct {
+	suite.Suite
+	logger        tools.Logger
+	successClient *mocks.RateClient
+	failClient *mocks.RateClient
+}
+
+func (t *ChainTestSuite) SetupSuite() {
+	l := zap.NewNop()
+	logger := l.Sugar()
+
+	t.logger = tools.NewZapLogger(logger)
+
+	mockRateClientS := mocks.NewRateClient(t.T())
+	mockRateClientS.On("GetDollarRate", mock.Anything).Return(expectedSuccessRate, nil)
+	t.successClient = mockRateClientS
+
+	mockRateClientF := mocks.NewRateClient(t.T())
+	mockRateClientF.On("GetDollarRate", mock.Anything).Return(expectedFailRate, errors.New("bank not available"))
+	t.failClient = mockRateClientF
+}
+
+func setChain(cf clients.RateClient, cs clients.RateClient) *clients.BaseChain {
+	clf := clients.NewBaseChain(cf)
+	cls := clients.NewBaseChain(cs)
+	clf.SetNext(cls)
+
+	return clf
+}
+
+func (t *ChainTestSuite) TestSuccessFirstClient() {
+	client := setChain(t.successClient, t.successClient)
+
+	rate, err := client.GetDollarRate(context.Background())
+
+	require.NoError(t.T(), err)
+	require.Equal(t.T(), expectedSuccessRate, rate)
+
+	t.successClient.AssertExpectations(t.T())
+}
+
+func (t *ChainTestSuite) TestSuccessSecondClient() {
+	client := setChain(t.failClient, t.successClient)
+
+	rate, err := client.GetDollarRate(context.Background())
+
+	require.NoError(t.T(), err)
+	require.Equal(t.T(), expectedSuccessRate, rate)
+
+	t.failClient.AssertExpectations(t.T())
+	t.successClient.AssertExpectations(t.T())
```

Comment:
Because expectations are unique for each separate test. Didn't get your point. Can you clarify?

Reply:
`AssertExpectations` is automatically called after all tests of suite have been runned. You can check it in generated constructor of mock.

There is approach when you setup mocks before each test and reset mocks after each test. In this way you need explicitly assert expectations after each test.

In your case, client mocks are not resetted during the tests. There is no need to call `AssertExpectations` explicitly in the end of specific test, because default check in the end of test suite will satisfy your needs.

My point is `explicit assert expectations in the end of each test seem to me redundant is this test suite`. If you are not agree with it, you can leave it as it is.

Reply:
Good point from @boredgusa.
Also i want to mention that your class doesn't make any `Expectations` cause you are using mock.Anything that skips it.
![image](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-RinaDish/assets/54763723/0295f367-4dcd-4e65-acbd-5783e69aacc1)

Reply:
As Darina mentioned in private messages, her idea is not relevant. But since you also mentioned it, I will ask again what you mean. Because I don't know how to make an AfterTest that is common for all tests check unique expectations. And for what purpose? Now each test has its own expectations, which is exactly what I wanted. Can you please clarify your idea? @MlPablo

Reply:
@MlPablo Done